src/kernel.ts

```diff
@@ -1,1069 +1,803 @@
+ */
+export
+interface IKernelFuture {
+  /**
+   * Dispose and unregister the future.
+     
+  dispose(): void;
+
+  /**
+   * Register a reply handler. Returns `this`.
+   */
+  onReply(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * Register an output handler. Returns `this`.
+   */
+  onOutput(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * Register a done handler. Returns `this`.
+   */
+  onDone(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * Register an input handler. Returns `this`.
+   */
+  onInput(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * The autoDispose behavior of the future.
+   *
+   * If True, it will self-dispose() after onDone() is called.
+   */
+  autoDispose: boolean;
+
+  /**
+   * Set when the message is done.  
+   */
+  isDone: boolean;
 
-    _on_error(error) {
-        /**
-         * Handle a failed AJAX request by logging the error message, and
-         * then optionally calling a provided callback.
-         *
-         * @function _on_error
-         * @param {function} error - callback
-         */
-        return (xhr, status, err) => {
-            utils.log_ajax_error(xhr, status, err);
-            if (error) {
-                error(xhr, status, err);
-            }
-        };
-    }
+}
 
-    _kernel_created(data): void {
-        /**
-         * Perform necessary tasks once the kernel has been started,
-         * including actually connecting to the kernel.
-         *
-         * @function _kernel_created
-         * @param {Object} data - information about the kernel including id
-         */
-        this.id = data.id;
-        this.kernel_url = utils.url_join_encode(this.kernel_service_url, this.id);
-        this.start_channels();
-    }
 
-    _kernel_connected(): void {
-        /**
-         * Perform necessary tasks once the connection to the kernel has
-         * been established. This includes requesting information about
-         * the kernel.
-         *
-         * @function _kernel_connected
-         */
-        events.trigger('kernel_connected.Kernel', {kernel: this});
-        // get kernel info so we know what state the kernel is in
-        var that = this;
-        this.kernel_info((reply?) => {
-            this.info_reply = reply.content;
-            events.trigger('kernel_ready.Kernel', {kernel: that});
+/**
+ * A Kernel class to communicate with the Python kernel. This
+ * should generally not be constructed directly, but be created
+ * by the `Session` object. Once created, this object should be
+ * used to communicate with the kernel.
+ */
+export
+class Kernel {
+
+  static statusChange = new Signal<Kernel, string>();
+
+  constructor(kernel_service_url: string, ws_url: string, name: string) {
+    this._id = null;
+    this._name = name;
+    this._ws = null;
+
+    this._kernelServiceUrl = kernel_service_url;
+    this._kernelUrl = null;
+    this._wsUrl = ws_url || utils.getBodyData('wsUrl');
```

Comment:
I think we should not encourage putting the websocket URL in the body element.  I think they did this so that the server-side generated html could encode the websocket URL.  Instead, I think that we should not rely on server-generated templates, and instead have this part of the json api in talking to the server.  So +1 in removing this dependency on data in the body element.